### PR TITLE
VL+CT Riemann Kernel

### PR DESCRIPTION
**Pull request by mabruzzo**
This PR depends on PR #139 (introduction of the CMake build-system).

As I describe below in more detail in [this comment](https://github.com/enzo-project/enzo-e/pull/145#issuecomment-1087722589), 
- the file changes introduced by this PR are actually 3056 insertions & 1296 deletions that affect 37 files (about half of what GitHub shows)
- the failing test in the CI can safely be ignored
In that comment, I also recommend how this PR can be reviewed (so that reviewers don't have to wait for PR#139 to be merged to be able to start reviewing this PR)

 
# Summary

This PR largely focuses on refactoring the Riemann Solver. In service of this refactor, it also involves some related changes to the implementation of `EnzoEFltArrayMap` & to the build-system.

The changes to the implementation of `EnzoEFltArrayMap` are described in the following spoiler tag:

<details>
  <summary>EnzoEFltArrayMap Changes</summary>

  * `EnzoEFltArrayMap` now internally stores its array data in a new template class called `CArrCollec` (which is implemented in the Array component of the Cello layer).

  * `CArrCollec` is effectively a union that supports 2 different approaches forstoring an ordered collection of 3D `CelloArray`s that share a fixed shape:

     1. A vector of `CelloArray`s (equivalent to an array of pointers)
     2. A single contiguous 4D `CelloArray`

  * Each approach has a nearly identical interface except that the latter can provides access to the underlying array (the way they are constructed is also somewhat different)

  * In the long-term, we probably just want to use a single backing implementation, but in the short-term, this substantially simplifies things. (This may merit further discussion - there might actually be an argument for supporting both approaches)

</details>

Changes to the Riemann Solvers include:

  * moving the Riemann Solver source/header files into the subdirectory `src/Enzo/EnzoRiemann`, and modifying the build system to compile the Riemann Solvers in a library separate from the rest of the Enzo-layer. The change in code organization follows conventions that were previously discussed (but never implemented). Compilation as a separate library **substantially** improves the speed with which you can develop the Riemann Solver code. (The template-heavy nature of this code previously would cause a rebuild of the entire Enzo-layer to be triggered whenever **any** implementation detail changed). This led to some changes to the PUP routines that appear a little messy, but can be removed once we move to our new restart strategy.
  * I refactored the way the Riemann Solvers are implemented to use a kernel-based approach that should be *very* compatible with porting to Kokkos.
  * I introduced the openmp-simd pragma to 2 for-loops in the Riemann solvers. Note that enabling openmp-simd compiler flags is totally separate from the rest of openmp (none of the other openmp pragmas are honored and the runtime library is *not* linked). I provide more details about related minor tweaks to the build system down below.
  * Various speed improvements (guided with Intel Advisor). This includes introducing the `FORCE_INLINE` macro which wraps compiler extensions to force function inlining.
  * I refactored the `EOSStructIdeal` struct (currently only used by the Riemann solvers) in anticipation of some upcoming changes that I plan to make to the equation of state machinery used throughout the VL+CT solver (these changes are briefly described below).
  * I replaced the HLLD Riemann Solver implementation with the implementation from Athena++. Their implementation is noticeably faster and is easier to read than the version I had ported from Enzo. I also added the Athena++ license to the repository for this code.

With the openmp-simd pragmas enabled, the VL+CT solver is at least 30% faster (I'll come back and update with more exact numbers)

My tweaks to the build system (related to openmp-simd pragmas) are described in the following spoiler tag:
<details>
  <summary>Build System Tweaks</summary>

  * The build system continues to target a build without value-unsafe optimizations for floating point operations by default
  * While configuring a build (i.e. executing `cmake`) users who want to enable the openmp-simd pragmas should pass the `-D USE_SIMD = ON` pragmas.
  * When the openmp-simd pragmas are enabled the build will enable value-unsafe floating point optimizations (with `-ffast-math`). It will also pass any flags in `CONFIG_ARCH_FLAGS` to the compiler; these inform the compiler which vector instructions the binary can/should use. `CONFIG_ARCH_FLAGS` should generally be defined in the config files since there will be different preferences on different machines. If the flag is not defined an explanatory error message is printed (as an aside, when `USE_SIMD = OFF`, a lack of a `CONFIG_ARCH_FLAGS` definition won't cause any problems)
  * For the generic build files, `linux_gcc.cmake` & `linux_icc.cmake`, I set the `CONFIG_ARCH_FLAGS` variables to `-march=native` & `-xHost` in order to tell the compilers to target the instruction-set used by the cpu that is performing the build
  * For `frontera_icc.cmake`, I set `CONFIG_ARCH_FLAGS` to `-xCORE-AVX512` since that is recommended by the frontera user-guide.

</details>

# Future Plans

Going forward, I plan to:

  * Refactor how we handle the equation of state. I've got (what I think is) an elegant solution in mind for replacing the `EnzoEquationOfState` class hierarchy that's just used in the VL+CT with some new lighter-weight, extendible machinery that wraps structs like `EOSStructIdeal`. It's also my intention to make this into a `Physics` object so that we can at least start to centralize where the EOS-info is all stored. (Currently it's all scattered around)

  * It's my intention to look into isolating other pieces of the hydro-infrastructure (e.g. reconstructors) from the rest of the Enzo-layer

  * I would like to try to vectorize other pieces of the hydro-infrastructure (but I'll play that by ear. It might not make sense in light of the ongoing Kokkos work)

  * I've got some ideas for better interfacing between `CelloArray`s and the `Field` component from the `Cello` layer. Among other benefits, this would definitely help with removing the need for multiple data organization backends in `CArrCollec`

### CODE REVIEW

**Comment by mabruzzo** [comment]
This PR is complete and now ready for review. 

Because this builds off of PR #139, it looks like this PR includes twice as many changes as it actually does. ``git diff -M20% --shortstat cmake_new-output vlct-riemann-kernel`` indicates that there are really 3056 insertions and 1296 deletions.

It's definitely possible to review this PR asynchronously with PR #139. I would just wait to review the following changes until after PR #139 get's merged into main
```
 CMakeLists.txt                                                           |   6 +
 README.md                                                                |   1 +
 cmake/EnableOMPSIMD.cmake (new)                                          | 104 ++++++++++++
 config/frontera_icc.cmake                                                |   6 +
 config/linux_gcc.cmake                                                   |  21 ++-
 config/linux_icc.cmake                                                   |  10 +-
 input/vlct/run_MHD_linear_wave_test.py                                   |   2 +-
 input/vlct/run_MHD_shock_tube_test.py                                    |   6 +-
 input/vlct/run_dual_energy_cloud_test.py                                 |   4 +-
 src/Cello/CMakeLists.txt                                                 |   5 +
 src/Enzo/CMakeLists.txt                                                  |  10 +-
 src/Enzo/EnzoRiemann/CMakeLists.txt (new)                                |  42 +++++
```
In the meantime, I would just focus on the following changes:
```
 LICENSE_ATHENAPP (new)                                                   |  31 ++++
 doc/source/devel/array.rst                                               | 272 ++++++++++++++++++++++++++++++-
 doc/source/devel/hydro_infrastructure.rst                                | 525 ++++++++++++++++++++++++++++++++++--------------------------
 src/Cello/_array.hpp                                                     |   1 +
 src/Cello/array_CArrCollec.hpp (new)                                     | 360 +++++++++++++++++++++++++++++++++++++++++
 src/Cello/array_CelloArray.hpp                                           |  37 ++++-
 src/Cello/cello.hpp                                                      |   1 +
 src/Cello/cello_defines.hpp (new)                                        |  28 ++++
 src/Cello/test_CArrCollec.cpp (new)                                      | 444 ++++++++++++++++++++++++++++++++++++++++++++++++++
 src/Cello/test_CelloArray.cpp                                            | 113 +++++++++++++
 src/Enzo/{enzo_EnzoRiemann.cpp => EnzoRiemann/EnzoRiemann.cpp}           |  63 ++++++--
 src/Enzo/{enzo_EnzoRiemann.hpp => EnzoRiemann/EnzoRiemann.hpp}           |  91 +++++++----
 src/Enzo/{enzo_EnzoRiemannHLL.hpp => EnzoRiemann/EnzoRiemannHLL.hpp}     | 176 ++++++++++++++------
 src/Enzo/{enzo_EnzoRiemannHLLC.hpp => EnzoRiemann/EnzoRiemannHLLC.hpp}   |  99 ++++++++----
 src/Enzo/EnzoRiemann/EnzoRiemannHLLD.hpp (new)                           | 456 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/Enzo/{enzo_EnzoRiemannImpl.hpp => EnzoRiemann/EnzoRiemannImpl.hpp}   | 513 +++++++++++++++++++---------------------------------------
 src/Enzo/{enzo_EnzoRiemannLUT.hpp => EnzoRiemann/EnzoRiemannLUT.hpp}     |   9 +-
 src/Enzo/{enzo_EnzoRiemannUtils.hpp => EnzoRiemann/EnzoRiemannUtils.hpp} | 439 +++++++++++++++++++++++++++++++++-----------------
 src/Enzo/_enzo.hpp                                                       |  13 +-
 src/Enzo/enzo.ci                                                         |   9 --
 src/Enzo/enzo_EnzoEFltArrayMap.cpp                                       |  90 ++---------
 src/Enzo/enzo_EnzoEFltArrayMap.hpp                                       |  37 +++--
 src/Enzo/enzo_EnzoMethodMHDVlct.cpp                                      |   4 +-
 src/Enzo/enzo_EnzoMethodMHDVlct.hpp                                      |   3 +
 src/Enzo/enzo_EnzoRiemannHLLD.hpp (gone)                                 | 321 -------------------------------------
```

For reference, I generated these lists with `git diff -M20% --compact-summary cmake_new-output vlct-riemann-kernel`

As an aside, I just wanted to mention that neither `git-diff` (without the `-M20%` flag) nor GitHub properly shows the differencess for the following files correctly: 
- `src/Enzo/EnzoRiemman/EnzoRiemann.cpp`
- `src/Enzo/EnzoRiemman/EnzoRiemannImpl.hpp`
- `src/Enzo/EnzoRiemman/EnzoRiemannUtils.hpp`

Since this PR introduced substantial changes after the files were moved, the file changes shown as though these are brand-new files (This is particularly annoying since I went out of my to isolate the file renaming to a separate commit).

**Review by gregbryan — APPROVED** [review]
Looks good to me (a few typos).

**Comment by mabruzzo** [comment]
@peeples When you have a chance, would you mind assigning a second reviewer?

**Pull request merged**
